velox/functions/lib/tests/TDigestTest.cpp

```diff
@@ -0,0 +1,384 @@
+    0.3000, 0.4000, 0.5000, 0.6000,  0.7000, 0.8000, 0.9000,
+    0.9500, 0.9600, 0.9700, 0.9800,  0.9999,
+};
+
+void checkQuantiles(
+    folly::Range<const double*> values,
+    const TDigest<>& digest) {
+  VELOX_CHECK(std::is_sorted(values.begin(), values.end()));
+  auto sum = std::accumulate(values.begin(), values.end(), 0.0);
+  ASSERT_NEAR(digest.sum(), sum, kSumError);
+  for (auto q : kQuantiles) {
+    auto v = digest.estimateQuantile(q);
+    ASSERT_LE(values.front(), v);
+    ASSERT_LE(v, values.back());
+    auto hi = std::lower_bound(values.begin(), values.end(), v);
+    auto lo = hi;
+    while (lo != values.begin() && v > *lo) {
+      --lo;
+    }
+    while (std::next(hi) != values.end() && *hi == *std::next(hi)) {
+      ++hi;
+    }
+    auto l = (lo - values.begin()) / (values.size() - 1.0);
+    auto r = (hi - values.begin()) / (values.size() - 1.0);
+    if (q < l) {
+      ASSERT_NEAR(l, q, kRankError);
+    } else if (q > r) {
+      ASSERT_NEAR(r, q, kRankError);
+    }
+  }
+}
+
+#define CHECK_QUANTILES(_values, _digest) \
+  do {                                    \
+    SCOPED_TRACE("CHECK_QUANTILES");      \
+    checkQuantiles((_values), (_digest)); \
+  } while (false)
+
+std::string decodeBase64(std::string_view input) {
+  std::string decoded(folly::base64DecodedSize(input), '\0');
+  folly::base64Decode(input, decoded.data());
+  return decoded;
+}
+
+TEST(TDigestTest, addElementsInOrder) {
+  constexpr int N = 1e6;
+  TDigest digest;
+  ASSERT_EQ(digest.compression(), tdigest::kDefaultCompression);
+  std::vector<int16_t> positions;
+  for (int i = 0; i < N; ++i) {
+    digest.add(positions, i);
+  }
+  digest.compress(positions);
+  ASSERT_NEAR(digest.sum(), 1.0 * N * (N - 1) / 2, kSumError);
+  for (auto q : kQuantiles) {
+    auto v = digest.estimateQuantile(q);
+    ASSERT_NEAR(v / (N - 1), q, kRankError);
+  }
+}
+
+TEST(TDigestTest, addElementsRandomized) {
+  constexpr int N = 1e5;
+  double values[N];
+  TDigest digest;
+  std::vector<int16_t> positions;
+  std::default_random_engine gen(common::testutil::getRandomSeed(42));
+  std::uniform_real_distribution<> dist;
+  for (int i = 0; i < N; ++i) {
+    auto v = dist(gen);
+    digest.add(positions, v);
+    values[i] = v;
+  }
+  digest.compress(positions);
+  std::sort(std::begin(values), std::end(values));
+  CHECK_QUANTILES(folly::Range(values, N), digest);
+}
+
+TEST(TDigestTest, fewElements) {
+  TDigest digest;
+  std::vector<int16_t> positions;
+  digest.compress(positions);
+  ASSERT_EQ(digest.sum(), 0);
+  for (auto q : kQuantiles) {
+    ASSERT_TRUE(std::isnan(digest.estimateQuantile(q)));
+  }
+  digest.add(positions, 1.0);
+  digest.compress(positions);
+  ASSERT_EQ(digest.sum(), 1);
+  for (auto q : kQuantiles) {
+    ASSERT_EQ(digest.estimateQuantile(q), 1.0);
+  }
+}
+
+// IMPORTANT: All these errors cannot be caught by TRY in Presto, so we should
+// not make them user errors.  If in another engine these are catchable errors,
```

Comment:
Sounds like a use case of VELOX_CHECK_UNSUPPORTED_INPUT_UNCATCHABLE()?

Reply:
Presto also throw exception in this case, so we can just throw runtime exception